manual/node_pattern.md

```diff
@@ -0,0 +1,270 @@
+# Node Pattern
+
+Node pattern is a small language to help find specific nodes in
+the Abstract Syntax Tree using a simple string.
+
+It reminds the simplicity of regular expressions but used to find specific
+nodes of ruby code.
+
+## Simple match
+
+A simple integer like `1` represents `(int 1)` in the AST.
+
+```
+$ ruby-parse -e '1'
+(int 1)
+```
+
+- `int` will match exactly the node
+- `(int 1)` will match precisely the node
+
+## `_` for any value
+
+`_` will check if there's something present in the specific position, no matter the
+value:
+
+- `(int _)` will match any number
+- `(int _ _)` will not match because `int` types have just one child that
+  contains the value.
+
+The compiler translates the pattern `(int _)` into something like:
+
+```ruby
+node.int_type? && node.children.size == 1
+```
+
+while `(int _ _)` will be translated to something like:
+
+```ruby
+node.int_type? && node.children.size == 2
+```
+
+## `...` for something else
+
+While `_` limits the size, `...` make it open for more children.
+
+And the example `(int ...)` will be translated into something like:
+
+```ruby
+node.int_type? && node.children.size >= 1
+```
+
+And it matches with any number as well.
+
+It's useful when you want to check some variable internal nodes but with a
+final with the same results. For example, let's use a classic `something.save`.
+
+We can also have `person.save` or `person.address.save`, and we want to match
+both. So, let's check how it looks like in the AST:
+
+```
+$ ruby-parse -e 'person.save'
+(send (send nil :person) :save)
+```
+
+The first case can be addressed with an expression like:
+
+```
+(send (send nil? _) :save)
+```
+
+But if it contains `.address` in the middle:
+
+```
+$ruby-parse -e 'person.address.save'
+(send
+  (send
+    (send nil :person) :address) :save)
+```
+
+The expression will not match, and it can be `a.b.c.d.save`.
+
+In such cases you can use `...` to match both cases:
+
+```
+(send ... :save)
+```
+
+## `{}` for any `<expression>`
+
+Lets make it a bit more complex and introduce floats:
+
+```
+$ ruby-parse -e '1'
+(int 1)
+$ ruby-parse -e '1.0'
+(float 1.0)
+```
+
+- `({int float} _)` - int or float types, no matter the value
+
+## `$` for captures
+
+You can capture elements or nodes along with your search, prefixing the expression
+with `$`. For example, in a tuple like `(int 1)` with a search like `(int $_)`,
+you can catch element if the search matches.
+
+You can also capture multiple things like:
+
+```
+(${int float} $_)
+```
+
+It will bring an array with `[:int, 1]` for a tuple like (int 1).
+The tuple can be captured using the `$` before the open parens:
+
+```
+$({int float} _)
+```
+
+## predicate methods
+
+You can also use methods that end with question mark `?`, to ask for some
+method in the node. Example:
+
+- `int_type?` can be used herein replacement of `(int _)`
```

Comment:
It shouldn't. What it can is a side effect of how it's implemented internally, e.g. I assume that Integer AST node class responds to `int_type?`, and as the result, `int` matches, not the other way around.

Reply:
Thing is, `int_type?` is a method of AST, while `odd?` is a method called on that object.

Reply:
Yes. Exactly. I think we need to emphasize it in somehow.